### PR TITLE
Feature/fgrgbwm 441

### PR DESCRIPTION
**Pull request by MatthewWaanders**

### CODE REVIEW

File: drivers/FGRGBWM-441/device.js [hunk 1]
```diff
@@ -1,102 +1,656 @@
...
+                } catch (err) {
+                    return Promise.reject(err);
+                }
+            }
+            if (args.animation === '11') {
+                args.animation = Math.round(Math.random() * (10-6) + 6);
+            }
+
+            try {
+                this.configurationSet({
```
Comment (RobinBol) [inline drivers/FGRGBWM-441/device.js]: Not awaiting and returning the result? What if the set fails?

File: drivers/FGRGBWM-441/device.js [hunk 2]
```diff
@@ -1,102 +1,656 @@
...
+                newInputConfig.input3 !== this.realInputConfigs.input3 ||
+                newInputConfig.input4 !== this.realInputConfigs.input4) {
+                    this.realInputConfigs = newInputConfig;
+                    return true;
+        }
+        return false;
+    }
+
+    async _sendColor(value, multiChannel) {
+        this.node.MultiChannelNodes[multiChannel].CommandClass.COMMAND_CLASS_SWITCH_MULTILEVEL.SWITCH_MULTILEVEL_SET({Value: value}, (err, result) => {
```
Comment (RobinBol) [inline drivers/FGRGBWM-441/device.js]: This doesn't work, did you test this? Either return the promise returned from the `SWITCH_MULTILEVEL_SET` or wrap a promise around the whole method call.

File: app.json [hunk 3]
```diff
@@ -5375,608 +5328,728 @@
...
+          4096,
+          8192,
+          12288,
+          16384
+        ],
+        "zwaveAllianceProductId": 1054,
+        "learnmode": {
+          "image": "/drivers/FGRGBWM-441/assets/learnmode.svg",
+          "instruction": {
+            "en": "Press the button on your Fibaro RGBW Controller three times",
```
Comment (RobinBol) [inline app.json]: Punctuation.

File: app.json [hunk 4]
```diff
@@ -5375,608 +5328,728 @@
...
+              "type": "number",
               "label": {
-                "en": "Blue LED",
-                "nl": "Blauwe LED"
-              }
+                "en": "Maximum brightness",
+                "nl": "Maximale helderheid"
+              },
+              "hint": {
+                "en": "Sets the maximum brightness the module will output. Range: 3 - 255",
```
Comment (RobinBol) [inline app.json]: Range can be removed

File: app.json [hunk 5]
```diff
@@ -5375,608 +5328,728 @@
...
               "label": {
-                "en": "White LED",
-                "nl": "Witte LED"
-              }
-            },
+                "en": "Minimum brightness",
+                "nl": "Minimale helderheid"
+              },
+              "hint": {
+                "en": "Sets the minimum brightness of the module. Range: 2 - 254",
```
Comment (RobinBol) [inline app.json]: Range again, also check other settings

File: app.json [hunk 6]
```diff
@@ -5375,608 +5328,728 @@
...
+              "type": "number",
               "label": {
-                "en": "Momentary switch",
-                "nl": "Kortstondige schakelaar"
-              }
+                "en": "Transition step time (mode 1)",
+                "nl": "Transitie stap tijd (mode 1)"
+              },
+              "hint": {
+                "en": "Sets the time (in milli-seconds) that each step takes when making a transition. Range: 0 (immediate), 1 - 60000",
```
Comment (RobinBol) [inline app.json]: miliseconds/miliseconden*

File: app.json [hunk 7]
```diff
@@ -5375,608 +5328,728 @@
...
+                "min": 0,
+                "max": 2.54,
+                "step": 0.01
+              },
+              "value": 0.1
             },
             {
-              "id": "huee27v3",
+              "id": "save_state",
+              "type": "checkbox",
```
Comment (RobinBol) [inline app.json]: No zwave object?

File: app.json [hunk 8]
```diff
@@ -5375,608 +5328,728 @@
...
+              },
+              "attr": {
+                "min": 0,
+                "max": 65534
+              },
+              "value": 30
             },
             {
-              "id": "huestrip",
+              "id": "kwh_threshold",
```
Comment (RobinBol) [inline app.json]: No zwave object?

File: drivers/FGRGBWM-441/device.js [hunk 9]
```diff
@@ -1,102 +1,656 @@
...
+                    red = (rgb.red / 255) * 99;
+                    green = (rgb.green / 255) * 99;
+                    blue = (rgb.blue / 255) * 99;
+                }
+            }
 
-	_setColor(rgb, duration) {
-		const promises = [];
+            try {
+                await this._sendColor(red, 2);
```
[the remark this answers is not in the document]
Comment (MatthewWaanders) [reply drivers/FGRGBWM-441/device.js]: We could send them in parallel, I've done it this way to assure each value was sent correctly before setting it as the "CurrentRGB" color

File: drivers/FGRGBWM-441/device.js [hunk 10]
```diff
@@ -1,102 +1,656 @@
...
+                blue = (rgbw.blue / 255) * 99;
+                white = (rgbw.white / 255) * 99;
+            } else {
+                red = (rgb.red / 255) * 99;
+                green = (rgb.green / 255) * 99;
+                blue = (rgb.blue / 255) * 99;
+            }
+
+            try {
+                await this._sendColor(red, 2);
```
[the remark this answers is not in the document]
Comment (MatthewWaanders) [reply drivers/FGRGBWM-441/device.js]: Fixed in the catch clause, not auto closed as marked on the wrong line

File: app.json (hunk 7 above)
Comment (MatthewWaanders) [reply app.json]: fixed, I believe the line doesn't match up with this one so not auto closed

File: app.json (hunk 8 above)
Comment (MatthewWaanders) [reply app.json]: fixed, I believe the line doesn't match up with this one so not auto closed

File: drivers/FGRGBWM-441/device.js [hunk 11]
```diff
@@ -222,20 +226,22 @@ class FibaroRGBWControllerDevice extends ZwaveDevice {
...
-        Registering report listeners for multichannel nodes
-        ================================================================
-         */
-        this.registerMultiChannelReportListener(2, 'SWITCH_MULTILEVEL', 'SWITCH_MULTILEVEL_REPORT', (report) => this._reportParser(report, 2));
-        this.registerMultiChannelReportListener(3, 'SWITCH_MULTILEVEL', 'SWITCH_MULTILEVEL_REPORT', (report) => this._reportParser(report, 3));
-        this.registerMultiChannelReportListener(4, 'SWITCH_MULTILEVEL', 'SWITCH_MULTILEVEL_REPORT', (report) => this._reportParser(report, 4));
-        this.registerMultiChannelReportListener(5, 'SWITCH_MULTILEVEL', 'SWITCH_MULTILEVEL_REPORT', (report) => this._reportParser(report, 5));
+        this.registerCapability('measure_voltage.input1', 'SWITCH_MULTILEVEL', {
+            multiChannelNodeId: 2,
+            reportParser: (report) => this._reportParser(report, 2)
```
Comment (RobinBol) [inline drivers/FGRGBWM-441/device.js]: What if the command class switch multilevel supported by the device is e.g. version 4 and the sys capability has a `reportParserV4` which is preferred over `reportParser`? I assume the `this._reportParser` method will never be called in that case.

**Pull request merged**